networkx/classes/coreviews.py

```diff
@@ -281,14 +281,16 @@ def __init__(self, d, NODE_OK):
         self.NODE_OK = NODE_OK
 
     def __len__(self):
-        return sum(1 for n in self)
+        if hasattr(self.NODE_OK, "length"):
+            return self.NODE_OK.length
+        if hasattr(self.NODE_OK, "nodes"):
+            return len(self.NODE_OK.nodes)
```

Comment:
Right -- I think this should work:
```suggestion
            return len(self.NODE_OK.nodes & self._atlas.keys())
```
Similarly for the case below too.